### PR TITLE
feat: account selector

### LINKED ISSUE #2598: Account dropdown wire up
### Overview

We should merge https://github.com/shapeshift/web/pull/2594 with the UI as is, and use this ticket to change the technical interface/get this component ready for usage.

### References and additional details

* Wrap the existing dumb component in a reusable smart component that can be used in features throughout the app

### Acceptance Criteria

* This component accepts `AccountId[]` as input prop
* Should include an optional? `AssetId` prop to filter for accounts containing that asset
* Also accepts an `onChange?: (accountId: AccountId) => unknown` callback prop

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

fancy pants account selector component

todo

* [ ] formatting, fonts seem kinda bland/jank as is

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #2598

## Risk

* none till we consume this component
* returning the wrong `AccountId` would be **bad**

### Engineering

* this components API is designed such that a consumer passes an `AssetId` and gets back an `AccountId` via the `onChange` callback
* turn on the multi account flag, set your `.env` log level and browser console to trace logging, and navigate to the accounts page
* see that there is a single log event emitted by the `onChange` callback for each chain supported by the connected wallet on the `/accounts` route load. testing with native wallet is preferred for obvious reasons.
* note that only BTC and LTC are rendered - these are the only chains that require user facing account selectors with the flag off before we ship multi account.
* switch the account type for UTXO chains - notice the correct `AccountId` is emitted once per change
* click the same account for an account based chain - notice no event is emitted.

### Operations

* does not require testing by operations - this component is only to be used by engineering behind feature flags for multi account work.

### CODE REVIEW

**Comment by 0xdef1cafe** [comment]
@reallybeard 
image
 halp

File: src/components/AccountDropdown/AccountDropdown.tsx [hunk 1]
```diff
@@ -7,54 +7,165 @@ import {
...
+
+  /**
+   * these chains already have multi account support via sending and receiving,
+   * and we need to use *and* render this new component while we retrofit the rest of the app
+   *
+   * the effectful logic above will still run for other chains, and return the first account
+   * via the onChange callback on mount, but nothing will be visually rendered
+   */
+  const existingMultiAccountChainIds = useMemo(() => [btcChainId, ltcChainId], [])
+  if (isMultiAccountsEnabled && !existingMultiAccountChainIds.includes(chainId)) return null
```
Comment (0xdef1cafe) [inline src/components/AccountDropdown/AccountDropdown.tsx:163]: uncomment this line to see a demo for all chains

File: src/state/slices/portfolioSlice/selectors.ts [hunk 2]
```diff
@@ -600,6 +604,20 @@ export const selectPortfolioAssets = createSelector(
...
+ * but can contain it
+ */
+export const selectPortfolioAccountIdsByAssetId = createCachedSelector(
+  selectPortfolioAccountIds,
+  selectAssetIdParamFromFilter,
+  (accountIds, assetId): AccountId[] => {
+    const { chainId } = fromAssetId(assetId)
+    return accountIds.filter(accountId => fromAccountId(accountId).chainId === chainId)
+  },
+)((_accountIds, { assetId }) => assetId ?? 'undefined')
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/selectors.ts:619]: really good example of how uncached selectors can interfere and cause rerenders across sibling components

File: src/components/AccountDropdown/AccountDropdown.tsx [hunk 3]
```diff
@@ -7,54 +7,165 @@ import {
...
+        <AccountSegment
+          key={accountNumber}
+          title={translate('accounts.accountNumber', { accountNumber })}
+          subtitle={''} // hide me until we have the option to "nickname" accounts
+        />
+        {accountIds.map((accountId, index) => (
+          <AccountChildOption
+            key={`${accountId}-${index}`}
+            title={makeTitle(accountId)}
+            cryptoBalance={fromBaseUnit(accountBalances[accountId][assetId], asset.precision)}
```
Comment (0xdef1cafe) [inline src/components/AccountDropdown/AccountDropdown.tsx:135]: this probably doesn't localize correctly

File: src/components/AccountDropdown/AccountDropdown.tsx [hunk 4]
```diff
@@ -7,54 +7,165 @@ import {
...
-  symbol: string
-}
-
 type AccountDropdownProps = {
-  onClick: (arg: string) => void
-  activeAccount: string | null
-  accounts: AccountItem[]
+  assetId: AssetId
   buttonProps?: ButtonProps
+  onChange?: (accountId: AccountId) => void
```
Comment (0xdef1cafe) [inline src/components/AccountDropdown/AccountDropdown.tsx:44]: is there ever a case where this should be optional?

Comment (0xdef1cafe) [reply src/components/AccountDropdown/AccountDropdown.tsx:44]: changed my own mind - this is a bug waiting to happen if it's optional

**Comment by reallybeard** [comment]
Made it so the list will always fit the content, and not wrap.
Screen Shot 2022-08-31 at 11 42 48 PM

File: src/components/AccountDropdown/AccountChildOption.tsx [hunk 5]
```diff
@@ -12,9 +12,17 @@ export const AccountChildOption = forwardRef<AccountChildRowProps, 'button'>(
   ({ title, cryptoBalance, symbol, children, ...props }, ref) => {
     const Component = (
       <MenuItemOption ref={ref} {...props}>
-        <Stack direction='row' justifyContent='space-between'>
-          <RawText>{title}</RawText>
-          <Amount.Crypto value={cryptoBalance} symbol={symbol} />
+        <Stack direction='row' justifyContent='space-between' fontSize='sm' spacing={4}>
+          <RawText fontWeight='bold' whiteSpace='nowrap'>
+            {title}
```
Comment (0xApotheosis) [inline src/components/AccountDropdown/AccountChildOption.tsx:17]: All caps is pretty rough here. Can we use title case, e.g. `Segwit Native`?

**Review by 0xApotheosis — COMMENTED** [review]
Interface checks out. Design/fonts still feel a little rough.

File: src/components/AccountDropdown/AccountChildOption.tsx (hunk 5 above)
Comment (gomesalexandre) [reply src/components/AccountDropdown/AccountChildOption.tsx:17]: Agreed that could get done as a low lift and would look much better e.g:

```typescript
'SEGWIT NATIVE'.split(' ').map(x => x.toLowerCase()).map(_.upperFirst).join(' ')
// => 'Segwit Native'

'SEGWIT'.split(' ').map(x => x.toLowerCase()).map(_.upperFirst).join(' ')
// => 'Segwit'
```

(can obviously be refactored with a nice LTR pipeline)

**Review by gomesalexandre — APPROVED** [review]
Tested according to the testing criteria:

- This isn't rendered when connected to an EVM-only wallet e.g MetaMask
- The right `AccountId` gets `console.trace`d onChange - tested against the URL of the matching account's page
  - Selecting the same `AccountId` doesn't fire an `onChange` event a second time

Early stamping this as I'm conceptually happy with this initial implementation and this is only used internally and behind flag - I'm not so concerned about improvements as I'm concerned about the external interface, which looks sane as we've discussed it while pairing on this yesterday.

This will be reused in a lot of places as part of multi account, and the component API is at a sane place for this.

File: src/components/AccountDropdown/AccountDropdown.tsx (hunk 3 above)
Comment (0xdef1cafe) [reply src/components/AccountDropdown/AccountDropdown.tsx:148]: nvm it does in the child 😮‍💨

File: src/components/AccountDropdown/AccountChildOption.tsx (hunk 5 above)
Comment (0xdef1cafe) [reply src/components/AccountDropdown/AccountChildOption.tsx:17]: fixed at the root cause https://github.com/shapeshift/web/pull/2625/commits/6eaaca2d136c1f6ecb44a1680e6698fefeeba5d8

**Review by gomesalexandre — APPROVED** [review]
Retested according to the previous testing criteria, LGTM

**Pull request merged**